aptos-move/block-executor/src/scheduler_v2.rs

```diff
@@ -223,28 +242,152 @@ impl SchedulerV2 {
+            && !self.is_halted()
+            && self.txn_status[next_to_commit_idx as usize].is_executed()
+        {
+            self.queueing_commits_lock.arm();
+        }
+
+        self.queueing_commits_lock.unlock();
+    }
+
+    pub(crate) fn commit_hooks_try_lock(&self) -> bool {
+        self.queueing_commits_lock.try_lock()
+    }
+
+    // Should be called (i.e. in a while loop until None) after a successful should_perform_commit_hooks
+    // call. Completing the hooks should be followed by a commit_hook_performed call.
+    pub(crate) fn try_get_sequential_commit_hook(
+        &self,
+    ) -> Result<Option<(TxnIndex, Incarnation)>, PanicError> {
+        // Relaxed ordering due to armed lock acq-rel.
+        let next_to_commit_idx = self.next_to_commit_idx.load(Ordering::Relaxed) as usize;
+
+        assert!(next_to_commit_idx <= self.num_txns as usize);
+        if self.is_halted() || next_to_commit_idx == self.num_txns as usize {
+            // All sequential commit hooks are already dispatched.
+            return Ok(None);
+        }
+
+        if next_to_commit_idx > 0
+            && self.committed_marker[next_to_commit_idx - 1].load(Ordering::Relaxed) != COMMITTED
+        {
+            // Must wait for the previous hook to complete.
+            return Ok(None);
+        }
+
+        if self.txn_status[next_to_commit_idx as usize].is_executed() {
+            // All prior transactions are committed and the latest incarnation of the transaction
+            // at next_to_commit_idx has finished but has not been aborted. If any of its read was
+            // incorrect, it would have been invalidated by the respective transaction's last
+            // (committed) (re-)execution, and led to an abort in the corresponding finish execution
+            // (which, inductively, must occur before the transaction is committed). Hence, it
+            // must also be safe to commit the current transaction.
+
+            if self.committed_marker[next_to_commit_idx as usize]
+                .swap(PENDING_COMMIT_HOOK, Ordering::Relaxed)
+                != NOT_COMMITTED
+            {
+                return Err(code_invariant_error(format!(
+                    "Marking {} as PENDING_COMMIT_HOOK, but previous marker != NOT_COMMITTED",
+                    self.post_commit_processing_queue.len()
+                )));
+            }
+
+            // Transaction might be executed despite a stalled status, e.g. intentionally,
+            // due to having a high priority, or simply if it gets stalled after the
+            // execution task is picked up. Hence, a transaction can be committed while
+            // stalled, and in this case its aborted dependencies need to be unstalled.
+            let mut aborted_dependency_guard =
+                self.aborted_dependencies[next_to_commit_idx as usize].lock();
+            let mut propagation_queue = BTreeSet::new();
+            if aborted_dependency_guard.is_stalled {
```

Comment:
But the transaction the stalled this txn is already committed. Does not it mean that it should have already unstall this tnx?

Reply:
no because it might have ignored stall to execute. this is assumes we allow it: stall is just a signal / hint, we can ignore it for super high priority txns for example or who knows some specific cases

Reply:
Fixed this - no deadlocks and tests pass with just normal recursive adding removing. 
Remembering just in case because I accidentally squashed the commits:
We decided not to proactively remove stall in commit, but wait for it to drain out. Should make more things visible during development and @sasha8 thinks it's also more intuitive.